gh-128715: Expose ctypes.CField, with info attributes

encukou opened the pull request:
Expose the type of struct/union field descriptors (`_ctypes._CField`) as `ctypes.CField`. (This is mainly to make it easier to refer to it -- for typing and documentation. Creating `CField`s manually is not supported.)

Add attributes for easier introspection:
- `name` & `type`, as defined in `_fields_`
- `byte_size` & `byte_offset`, specify the chunk of memory to read/write
- `bit_size` & `bit_offset`, which specify the shift & mask for bitfields
- `is_bitfield` & `is_anonymous` booleans

The existing `offset` remains, as an alias for `byte_offset`.
The existing `size` is unchanged. Usually it is the same as `byte_size`, but for bitfields, it contains a bit-packed combination of `bit_size` & `bit_offset`.

Update the `repr()` of `CField`.


Use the same values internally as well. (Except `bit_size`, which might overflow `Py_ssize_t` for very large types. Instead, in C use `bitfield_size`, which is 0 for non-bitfields and thus serves as `is_bitfield` flag. Different name used clarity.)
Old names are removed from the C implementation to ensure a clean transition.
For simplicity, I keep `byte_size` in `CFieldObject` for now, even though it's redundant: it's the size of the underlying type.


Add a generic test that ensures the values are consistent, and that we don't have overlapping fields in structs. Use this check throughout `test_ctypes`, wherever a potentially interesting Structure or Union is created. (Tests for how simple structs *behave* after they're created don't need the checks, but I erred on the side of adding checks.)


Lift the restriction on maximum field size that was temporarily added in GH-126938. The max size is now `Py_ssize_t`.


This PR does not yet touch `cfield.c` getters & setters: the bit-packed “size” argument is computed and passed to those.



<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-128715 -->
* Issue: gh-128715
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--128950.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

encukou committed:
Restore max field size to sys.maxsize, as in Python 3.13 & below

encukou committed:
PyCField: Split out bit/byte sizes/offsets.

encukou committed:
Expose CField

encukou committed:
Add generic checks for all the test structs/unions

encukou committed:
More testing

encukou committed:
Tests: import CField from ctypes

encukou committed:
Clarify bit_offset

encukou committed:
Add a blurb

encukou committed:
Regen

encukou committed:
Merge in the main branch

encukou committed:
include <stdbool.h> in the common header

encukou committed:
Explicit casts

encukou committed:
Remove problematic assert

encukou committed:
Merge in the main branch

encukou committed:
Add a test for the new info, and fix 'name' for nested anonymous structs

picnixz commented on Lib/ctypes/_layout.py:
```diff
@@ -294,10 +296,16 @@ def get_layout(cls, input_fields, is_struct, base):
+
+            # Do not use CField outside ctypes, yet.
+            # The constructor is internal API and may change without warning.
+            _internal_use=True,
```
I think this is something that pytest does a lot but how about naming it `_cpython=True` to stress that it's tied to CPython implementation?

encukou replied:
Not really -- if there's another implementation of `ctypes` I'd prefer that they use `_internal_use` too :)

picnixz commented on Lib/test/test_ctypes/_support.py:
```diff
@@ -22,3 +18,115 @@ class _X(Structure):
+                    expected_size = (field.bit_size << 16) + offset_for_size
+                    self.assertEqual(field.size, expected_size)
+                else:
+                    # size == byte_size
```
I don't know if you used `===` above to not uncomment by mistake, so maybe `size === byte_size` as well? (or maybe the === was a typo)

encukou replied:
Wow, good eye :)
Above it's not a guard against uncommenting; I meant something like `≡` -- they're *always* the same. Here, the `else` case is where they're equal.

(The comments don't communicate that well, but, it doesn't really matter. The next person can make them consistent if they prefer that.)

picnixz commented on Lib/test/test_ctypes/_support.py:
```diff
@@ -22,3 +18,115 @@ class _X(Structure):
+                if not is_struct:
+                    # union fields may overlap
+                    next_bit = 0
+            except Exception as e:
```
Off-topic but this makes me wonder whether a re-raising context would be interesting:

```py
with add_note_on_failure(f'while checking field {name!r}: {field}'):
	...
```

I don't know if we're adding notes a lot in the test folder though so maybe not that useful. OTOH, this could be a `subTest` instead.

encukou replied:
Oh, you're right! Thanks!
I seem to recall running into some kind of trouble with `subTest` in a mixin, but I can't reproduce that any more.

picnixz commented on Modules/_ctypes/cfield.c:
```diff
@@ -56,51 +55,42 @@ Py_ssize_t LOW_BIT(Py_ssize_t offset);
+    if (!_internal_use) {
+        // Do not instantiate outside ctypes, yet.
+        // The constructor is internal API and may change without warning.
+        PyErr_Format(PyExc_TypeError, "Cannot create %T object.", type);
```
```suggestion
        PyErr_Format(PyExc_TypeError, "cannot create %T object", type);
```

Considering all exceptions messages by a lowercase letter and don't end with a period, let's be consistent here as well

picnixz commented on Modules/_ctypes/cfield.c:
```diff
@@ -144,11 +129,69 @@ PyCField_new_impl(PyTypeObject *type, PyObject *name, PyObject *proto,
+                        "bit field %R size too large, got %zd",
+                        name, byte_size);
+        }
+        bitfield_size = PyLong_AsSsize_t(bit_size_obj);
```
Note for myself: because `bit_size_obj` should be small, we would use a fast path so this call would be efficient. However, if it's too large, then it will be slower, but we don't care as we will raise an exception.

encukou replied:
This is class creation. I'm leaving optimizations to PGO :)

picnixz commented on Modules/_ctypes/cfield.c:
```diff
@@ -144,11 +129,69 @@ PyCField_new_impl(PyTypeObject *type, PyObject *name, PyObject *proto,
+        }
+    }
+
+    self = _CFieldObject_CAST(type->tp_alloc(type, 0));
```
I think I left out this macro usage because, if in the future, we add assertions in the macro instead of using it for a raw pointer cast, then we would be doing something un-necessary. But it's up to you.

encukou replied:
This doesn't need to be super-fast, let's do more assertions ;)

picnixz commented on Modules/_ctypes/cfield.c:
```diff
@@ -144,11 +129,69 @@ PyCField_new_impl(PyTypeObject *type, PyObject *name, PyObject *proto,
+        self->name = Py_NewRef(name);
+    } else {
+        self->name = PyObject_Str(name);
+        if (!self->name) {
```
```suggestion
    }
    else {
        self->name = PyObject_Str(name);
        if (!self->name) {
```

picnixz commented on Modules/_ctypes/cfield.c:
```diff
@@ -232,25 +287,101 @@ PyCField_get(PyObject *op, PyObject *inst, PyTypeObject *type)
+    }
+    PyObject *eight = PyLong_FromLong(8);
+    if (!eight) {
+        return NULL;
```
```suggestion
        Py_DECREF(byte_size_obj);
        return NULL;
```

picnixz commented on Modules/_ctypes/cfield.c:
```diff
@@ -232,25 +287,101 @@ PyCField_get(PyObject *op, PyObject *inst, PyTypeObject *type)
+    if (!eight) {
+        return NULL;
+    }
+    return PyNumber_Multiply(byte_size_obj, eight);
```
Don't you also need to decref eight and byte_size_obj?

picnixz commented on Modules/_ctypes/cfield.c:
```diff
@@ -232,25 +287,101 @@ PyCField_get(PyObject *op, PyObject *inst, PyTypeObject *type)
+
+static PyMemberDef PyCField_members[] = {
+    { "name",
+        .type=Py_T_OBJECT_EX,
```
Nit but I think we write `.member = ...` with surrounding spaces around `=` for clarity.

picnixz commented on Modules/_ctypes/cfield.c:
```diff
@@ -144,11 +129,69 @@ PyCField_new_impl(PyTypeObject *type, PyObject *name, PyObject *proto,
+    if (PyUnicode_CheckExact(name)) {
+        self->name = Py_NewRef(name);
+    } else {
+        self->name = PyObject_Str(name);
```
IIRC, in `CFieldObject`, `name` is documented as being an *exact* unicode object but this is not necessarily the case as `PyObject_Str` may return a non-exact unicode object.

encukou replied:
Thanks for catching these!

encukou committed:
Use subTest

encukou committed:
Use PyUnicode_FromObject to get an exact PyUnicode

encukou committed:
Normalize exception message

encukou committed:
Fix refcounting

encukou committed:
Add pretty spaces

encukou committed:
Merge in the main branch

encukou committed:
Fix bit-packed size test for big-endian machines

encukou committed:
Remove `size` from _layout.py

encukou committed:
Fix bit_offset for big-endian structs (where bitfields are laid out from the big end)

encukou committed:
Merge in the main branch

encukou committed:
Name the magic constant

encukou committed:
Remove unused variable

encukou committed:
Remove an unacceptable blank line

encukou committed:
Update documentation for tp_basicsize & tp_itemsize

- Add alignment requirement
- Mention that ob_size is unreliable if you don't control it
- Add some links for context
- basicsize should include the base type in generaly not just PyObject

This adds a “by-the-way” link to `PyObject_New`, which shouldn't be
used for GC types. In order to be comfortable linking to it, I also
add a link to `PyObject_GC_New` from its docs. And the same for
`*Var` variants, while I'm here.

encukou committed:
Merge in the main branch

encukou committed:
Skip "is in" test for bitfields of underaligned types (bug filed)

picnixz commented on Lib/test/test_ctypes/_support.py:
```diff
@@ -22,3 +20,132 @@ class _X(Structure):
+        cls_size = ctypes.sizeof(cls)
+        for name, requested_type, *rest_of_tuple in cls._fields_:
+            field = getattr(cls, name)
+            with self.subTest(field=field):
```
To ease debugging maybe add "name=name" as well

picnixz commented on Lib/test/test_ctypes/test_aligned_structures.py:
```diff
@@ -5,9 +5,9 @@
 
-
-class TestAlignedStructures(unittest.TestCase):
+class TestAlignedStructures(unittest.TestCase, StructCheckMixin):
```
I am not entirely sure but shouldn't mixin classes appear before non-mixin ones?

encukou replied:
If the methods/attributes don't overlap, it doesn't matter.

picnixz commented on Modules/_ctypes/cfield.c:
```diff
@@ -56,51 +55,41 @@ Py_ssize_t LOW_BIT(Py_ssize_t offset);
 @classmethod
 _ctypes.CField.__new__ as PyCField_new
 
+    *
```
Maybe the name could be non-keyword?

encukou replied:
Let's do that when it's exposed as a public API.

picnixz commented on Modules/_ctypes/cfield.c:
```diff
@@ -144,11 +128,65 @@ PyCField_new_impl(PyTypeObject *type, PyObject *name, PyObject *proto,
+        if ((bitfield_size <= 0) || (bitfield_size > 255)) {
+            if (!PyErr_Occurred()) {
+                PyErr_Format(PyExc_ValueError,
+                            "bit size of field %R out of range, got %zd",
```
I think it's not aligned vertically

picnixz commented on Modules/_ctypes/cfield.c:
```diff
@@ -144,11 +128,65 @@ PyCField_new_impl(PyTypeObject *type, PyObject *name, PyObject *proto,
+            // so they're limited to about 8 bytes.
+            // This check is here to avoid overflow in later checks.
+            PyErr_Format(PyExc_ValueError,
+                        "bit field %R size too large, got %zd",
```
ditto

picnixz commented on Modules/_ctypes/cfield.c:
```diff
@@ -144,11 +128,65 @@ PyCField_new_impl(PyTypeObject *type, PyObject *name, PyObject *proto,
+                        name, byte_size);
+        }
+        bitfield_size = PyLong_AsSsize_t(bit_size_obj);
+        if ((bitfield_size <= 0) || (bitfield_size > 255)) {
```
You technically have redundant parentheses. Were they deliberate?

encukou replied:
Yeah, I don't trust C's  operator precedence :)

picnixz commented on Modules/_ctypes/cfield.c:
```diff
@@ -144,11 +128,65 @@ PyCField_new_impl(PyTypeObject *type, PyObject *name, PyObject *proto,
+    if (!self->name) {
+        goto error;
     }
+    assert (PyUnicode_CheckExact(self->name));
```
```suggestion
    assert(PyUnicode_CheckExact(self->name));
```

picnixz commented on Modules/_ctypes/cfield.c:
```diff
@@ -144,11 +128,65 @@ PyCField_new_impl(PyTypeObject *type, PyObject *name, PyObject *proto,
+            PyErr_Format(
+                PyExc_ValueError,
+                "field %R: bit_offset must be specified if bit_size is",
+                name);
```
Don't you have a missing goto error?

picnixz commented on Modules/_ctypes/cfield.c:
```diff
@@ -144,11 +128,65 @@ PyCField_new_impl(PyTypeObject *type, PyObject *name, PyObject *proto,
+            // This check is here to avoid overflow in later checks.
+            PyErr_Format(PyExc_ValueError,
+                        "bit field %R size too large, got %zd",
+                        name, byte_size);
```
Don't you have a missing goto error?

picnixz commented on Modules/_ctypes/cfield.c:
```diff
@@ -144,11 +128,65 @@ PyCField_new_impl(PyTypeObject *type, PyObject *name, PyObject *proto,
+                PyExc_ValueError,
+                "bit field %R overflows its type (%zd + %zd >= %zd)",
+                name, bit_offset, byte_size*8);
+                goto error;
```
I think the goto is indented one level too much

picnixz commented on Modules/_ctypes/cfield.c:
```diff
@@ -232,25 +279,110 @@ PyCField_get(PyObject *op, PyObject *inst, PyTypeObject *type)
 
 static PyObject *
-PyCField_get_size(PyObject *self, void *data)
+PyCField_is_anonymous(PyObject *self, void *data)
```
Since you're anyway modifying this line you can also replace void *data by `void *Py_UNUSED(closure)` if you want (I don't remember the exact macro name)

picnixz commented on Modules/_ctypes/cfield.c:
```diff
@@ -282,24 +414,27 @@ PyCField_dealloc(PyObject *self)
+            "<%T %R type=%s, ofs=%zd, bit_size=%zd, bit_offset=%zd>",
+            self,
+            field->name, tp_name, field->byte_offset,
+            (Py_ssize_t)field->bitfield_size,
+            (Py_ssize_t)field->bit_offset);
```
Is it usually preferrable to have a cast and use zd or directly use PRI8u?

encukou replied:
Generally, I don't know. `Py_ssize_t` seems more readable to me.

Here, it's an intentional (but somewhat obscure) choice -- these *are* sizes; they're just packed rather aggressively. Before there's a dedicated `CBitField` subclass, I don't want to waste memory on bitfield data.

picnixz reviewed (commented):
Some comments (hard to do a better review on a phone)

encukou committed:
Merge in the main branch

encukou committed:
Address review

encukou reviewed (commented):
Thank you for the review!

encukou committed:
One more alignment

sergey-miryanov commented on Modules/_ctypes/cfield.c:
```diff
@@ -110,17 +99,12 @@ PyCField_new_impl(PyTypeObject *type, PyObject *name, PyObject *proto,
                      "type of field %R must be a C type", self->name);
```
At this point `self` is NULL. I think should be just `name`.

picnixz replied:
We should convert `name` to an exact Unicode object first though but not bind it to `self` yet.

encukou replied:
D'oh, thanks!

`%R` on the input is fine for error messages -- the user is getting what they asked for.

sergey-miryanov commented on Modules/_ctypes/cfield.c:
```diff
@@ -144,11 +128,67 @@ PyCField_new_impl(PyTypeObject *type, PyObject *name, PyObject *proto,
+    if (!self) {
+        return NULL;
+    }
+    self->name = PyUnicode_FromObject(name);
```
@picnixz Eah, move this to the top and add a cleanup on error, if I get you right.

encukou committed:
Don't use `self` while it's NULL

picnixz reviewed (approved):
I don't think I've much more to say. My main concern was the unicode-exactness of the name which is now resolved so I'm fine.
